client/lib/html-toolbar/index.js

```diff
@@ -0,0 +1,228 @@
+/** @format */
+/**
+ * External dependencies
+ */
+import { Env } from 'tinymce/tinymce';
+import { noop, reduce } from 'lodash';
+
+/**
+ * Return an object containing the textarea content split into:
+ * - `before` selection
+ * - (`inner`) selection
+ * - `after` selection
+ *
+ * @param {Element} textarea The editor textarea.
+ * @returns {Object} The split textarea content.
+ */
+export const splitSelectedContent = textarea => {
+	const { selectionEnd, selectionStart, value } = textarea;
+	return {
+		before: value.substring( 0, selectionStart ),
+		inner: value.substring( selectionStart, selectionEnd ),
+		after: value.substring( selectionEnd, value.length ),
+	};
+};
+
+/**
+ * Set the cursor position after the insertion of new content into a textarea.
+ *
+ * @param {Element} textarea The editor textarea.
+ * @param {Number} selectionEnd The textarea.selectionEnd value BEFORE the content insertion.
+ * @param {Number} insertedContentLength The length of the inserted content.
+ */
+export const setCursorPosition = ( textarea, selectionEnd, insertedContentLength ) => {
+	textarea.selectionEnd = textarea.selectionStart = selectionEnd + insertedContentLength;
+};
+
+/**
+ * Insert content into a textarea.
+ *
+ * @param {Element} textarea The target textarea.
+ * @param {String} before The content before the selection.
+ * @param {String} content The new content to insert.
+ * @param {String} after The content after the selection.
+ * @param {Function} onInsert A function to call after the insertion.
+ */
+export const insertContent = ( textarea, before, content, after, onInsert = noop ) => {
+	const { selectionEnd, value } = textarea;
+	const newContent = before + content + after;
+
+	if ( Env.gecko ) {
+		// In Firefox, execCommand( 'insertText' ), needed to preserve the undo stack,
+		// always moves the cursor to the end of the content.
+		// A workaround involving manually setting the cursor position and inserting the editor content
+		// is needed to put the cursor back to the correct position.
+		textarea.value = newContent;
+		textarea.focus();
+		document.execCommand( 'insertText', false, content );
+		setCursorPosition( textarea, selectionEnd, newContent.length - value.length );
+		textarea.focus();
+		onInsert( newContent );
```

Comment:
Does anything bad happen if this is also the default action for Chrome?

Reply:
Not bad, just unnecessary.
On Chrome `insertText` triggers the textarea's `onChange`, so there is no need to manually update the component state.

(Actually, but I haven't tested it, there might even be a duplication of the content insertion in the undo history 🤔)